pootle/apps/pootle_fs/plugin.py

```diff
@@ -0,0 +1,344 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+import logging
+import os
+import shutil
+
+from django.contrib.auth import get_user_model
+from django.utils.functional import cached_property
+from django.utils.lru_cache import lru_cache
+
+from pootle.core.delegate import (
+    config, response as pootle_response, state as pootle_state)
+from pootle_store.models import Store
+from pootle_project.models import Project
+
+from .decorators import emits_state, responds_to_state
+from .delegate import fs_finder, fs_matcher, fs_resources
+from .signals import fs_pre_push, fs_post_push, fs_pre_pull, fs_post_pull
+
+
+logger = logging.getLogger(__name__)
+
+
+class Plugin(object):
+    """Base Plugin implementation"""
+
+    name = None
+
+    def __init__(self, project):
+        if not isinstance(project, Project):
+            raise TypeError(
+                "pootle_fs.Plugin expects a Project")
+        self.project = project
+
+    def __eq__(self, other):
+        return (
+            isinstance(other, self.__class__)
+            and self.project == other.project
+            and self.name == other.name)
+
+    def __str__(self):
+        return "<%s(%s)>" % (self.__class__.__name__, self.project)
+
+    @property
+    def is_cloned(self):
+        return os.path.exists(self.project.local_fs_path)
+
+    @property
+    def finder_class(self):
+        return fs_finder.get(self.__class__)
+
+    @property
+    def fs_url(self):
+        return self.project.config["pootle_fs.fs_url"]
+
+    @property
+    def response_class(self):
+        return pootle_response.get(self.state_class)
+
+    @cached_property
+    def matcher_class(self):
+        return fs_matcher.get(self.__class__)
+
+    @property
+    def store_fs_class(self):
+        from .models import StoreFS
+
+        return StoreFS
+
+    @cached_property
+    def matcher(self):
+        return self.matcher_class(self)
+
+    @cached_property
+    def pootle_user(self):
+        User = get_user_model()
+        username = config.get(
+            self.project.__class__,
+            instance=self.project,
+            key="pootle_fs.pootle_user")
+        if username:
+            try:
+                return User.objects.get(username=username)
+            except User.DoesNotExist:
+                logger.warning(
```

Comment:
this is for misconfigured user - its a pretty corner case, so i think logging that it cant use the configured user is appropriate.